src/riscv64/Gstep.c

```diff
@@ -0,0 +1,164 @@
+"Software"), to deal in the Software without restriction, including
+without limitation the rights to use, copy, modify, merge, publish,
+distribute, sublicense, and/or sell copies of the Software, and to
+permit persons to whom the Software is furnished to do so, subject to
+the following conditions:
+
+The above copyright notice and this permission notice shall be
+included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
+
+#include "unwind_i.h"
+#include "offsets.h"
+
+static int
+riscv64_handle_signal_frame (unw_cursor_t *cursor)
+{
+  int ret, i;
+  struct cursor *c = (struct cursor *) cursor;
+  unw_word_t sc_addr, sp, sp_addr = c->dwarf.cfa;
+  struct dwarf_loc sp_loc = DWARF_LOC (sp_addr, 0);
+
+  if ((ret = dwarf_get (&c->dwarf, sp_loc, &sp)) < 0)
+    return -UNW_EUNSPEC;
+
+  if (!unw_is_signal_frame (cursor))
+    return -UNW_EUNSPEC;
+
+  /* rt_sigframe contains the siginfo structure, the ucontext, and then
+     the trampoline. We store the mcontext inside ucontext as sigcontext_addr.
+  */
+  c->sigcontext_format = RISCV64_SCF_LINUX_RT_SIGFRAME;
+  c->sigcontext_addr = sp_addr + sizeof (siginfo_t) +
+    LINUX_UC_MCONTEXT_OFF;
+  c->sigcontext_sp = sp_addr;
+  c->sigcontext_pc = c->dwarf.ip;
+
+  /* Update the dwarf cursor.
+     Set the location of the registers to the corresponding addresses of the
+     uc_mcontext / sigcontext structure contents.  */
+
+#define  SC_REG_OFFSET(X)   (8 * X)
+
+  /* The PC is stored in place of X0 in sigcontext */
+  c->dwarf.loc[UNW_TDEP_IP] = DWARF_LOC (c->sigcontext_addr + SC_REG_OFFSET(UNW_RISCV64_X0), 0);
+
+  for (i = UNW_RISCV64_X1; i <= UNW_RISCV64_F31; i++)
+    {
+      c->dwarf.loc[i] = DWARF_LOC (c->sigcontext_addr + SC_REG_OFFSET(i), 0);
+    }
+
+  /* Set SP/CFA and PC/IP.  */
+  dwarf_get (&c->dwarf, c->dwarf.loc[UNW_TDEP_SP], &c->dwarf.cfa);
+  dwarf_get (&c->dwarf, c->dwarf.loc[UNW_TDEP_IP], &c->dwarf.ip);
+
+  return 1;
+}
+
+/*
+  Let's recognize PLT entries. As an example:
+
+  auipc t3, 0x1        0x00001e17
+  ld t3, 1904(t3)      0x770e3e03
+  jalr t1, t3          0x000e0367
+  nop                  0x00000013
+
+  The 0x1 and 1904 here are sample values for %pcrel_hi
+  and %pcrel_lo. We mask them out when checking the instructions.
+
+  For more information, see <https://github.com/riscv/riscv-elf-psabi-doc/blob/74ecf07bcebd0cb4bf3c39f3f9d96946cd6aba61/riscv-elf.md#program-linkage-table>.
+*/
+static int
+is_plt_entry (struct dwarf_cursor *c)
```

Comment:
This function is unused.

Reply:
Good catch. I added this while solving an unrelated issue. It shouldn't be necessary since we are not implementing fast unwind here, so let's remove it for now.